src/psyclone/dynamo0p3.py

```diff
@@ -3771,57 +3847,136 @@ def __init__(self, invoke, unique_psy_vars):
 
         # If we didn't have any inter-grid kernels but distributed memory
         # is enabled then we will still need a mesh object if we have one or
-        # more kernels that operate on cell-columns. We also require a mesh
+        # more kernels that operate on cell-columns or are doing redundant
+        # computation for a kernel that operates on dofs. Since the latter
+        # condition comes about through the application of a transformation,
+        # we don't yet know whether or not a mesh is required. Therefore,
+        # the only solution is to assume that a mesh object is required if
+        # distributed memory is enabled. We also require a mesh
         # object if any of the kernels require properties of either the
         # reference element or the mesh. (Colourmaps also require a mesh
         # object but that is handled in _colourmap_init().)
         if not _name_set:
-            if (requires_mesh or (Config.get().distributed_memory and
-                                  not invoke.operates_on_dofs_only)):
-                _name_set.add(self._schedule.symbol_table.find_or_create_tag(
-                    "mesh").name)
+            if requires_mesh or Config.get().distributed_memory:
```

Comment:
The `if not _name_set` means "have we already got the name of at least one mesh object?" if that's your first question? As to the second, you're quite right and `requires_mesh` is redundant and I've removed it.

Reply:
Thanks for simplifying. My first question was asking why we need to check whether we have the name of at least one mesh. I presumed it was really checking whether the name "mesh" had already been added. If that is the case, we don't care as it is a set so adding it will not result in a duplicate. I guess you are saying that we don't need to add "mesh" if *any* other mesh (possibly with a different name) has already been added. That is my understanding from your reply but I thought I'd double check. If so, could you update the comment above as this is not clear (to me) from the text.

Reply:
OK. I've reworded. Hopefully it's clearer now.

Reply:
Thanks.